examples/advanced/quantum_volume.py

```diff
@@ -136,7 +153,86 @@ def sample_heavy_set(circuit: cirq.Circuit,
+
+    Args:
+        circuit: The model circuit to compile.
+        device: The device to compile onto.
+        deconstruct: An optional function to deconstruct the model circuit's
+            gates down to the targer devices gate set.
+        optimize: An optional function to optimize the circuit's gates
+            after routing.
+
+    Returns: A tuple where the first value is the compiled circuit and the
+        second value is the final mapping from the model circuit to the compiled
+        circuit. The latter is necessary in order to preserve the measurement
+        order.
+
+    """
+    compiled_circuit = circuit.copy()
+    # Step 1: Unrolling. Descend into each gate's hierarchical definition and
+    # rewrite it in terms of lower-level gates that are in the target device
+    # gate set.
+    if deconstruct:
+        deconstruct(compiled_circuit)
+    # Step 2: Swap Mapping (Routing). Ensure the gates can actually operate on
+    # the target qubits given our topology.
+    swap_network = ccr.route_circuit(compiled_circuit,
+                                     ccr.xmon_device_to_graph(device),
+                                     algo_name='greedy')
+    compiled_circuit = swap_network.circuit
+    # Step 3: Optimize. The paper uses various optimization techniques - because
+    # Quantum Volume is intended to test those as well, we allow this to be
+    # passed in. This optimizer is not allowed to change the order of the
+    # qubits.
+    if optimize:
+        compiled_circuit = optimize(compiled_circuit)
+
+    return (compiled_circuit, swap_network.final_mapping())
+
+
+class QuantumVolumeResult:
+    """Stores all of the results and test information used when running the
+    quantum volume benchmark so it may be analyzed in detail afterwards.
+
+    """
+
+    def __init__(self):
+        # All of the model circuits used.
+        self.model_circuits: [cirq.Circuit] = []
+        # All of the heavy sets computed from the give model circuits.
+        self.heavy_sets: [List[int]] = []
+        # All of the model circuits after being compiled.
+        self.compiled_circuits: [cirq.Circuit] = []
+        # A map from each sampler index top a list of its HOG probabilities.
+        self.sampler_results: Dict[int, [float]] = defaultdict(list)
+
+    def _json_dict_(self):
+        return cirq.protocols.obj_to_dict_helper(self, [
+            'model_circuits', 'heavy_sets', 'compiled_circuits',
+            'sampler_results'
+        ])
+
+
+def main(*,
+         num_qubits: int,
+         depth: int,
+         num_repetitions: int,
+         seed: int,
+         device: cirq.google.xmon_device.XmonDevice,
+         samplers: List[cirq.Sampler],
+         deconstruct: PointOptimizer = None,
+         optimize: Callable[[cirq.Circuit], cirq.Circuit] = None
+        ) -> QuantumVolumeResult:
     """Run the quantum volume algorithm.
 
     The Quantum Volume benchmark is fairly straightforward. This algorithm will
```

Comment:
can we remove the line "The Quantum Volume benchmark is fairly straightforward"? At best it doesn't really add anything and at worst it could discourage a user who doesn't find it straightforward :)

Reply:
Haha yea, I still remember the lesson of "don't use the word 'obvious' in a paper" from one of my college profs... Looks like I forgot that here. Done.